### PR TITLE
drivers: dac: add mcp4728 driver

### PR DESCRIPTION
**Pull request by m-janus**
MCP4728 is a 12-bit, Quad Digital-to-Analog Converter with EEPROM Memory.
Controlled via I2C interface.

Signed-off-by: Marek Janus <marek.janus@grinn-global.com>

### CODE REVIEW

File: drivers/dac/dac_mcp4728.c [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+#include <zephyr.h>
+#include <drivers/dac.h>
+#include <drivers/i2c.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(dac_mcp4728, CONFIG_DAC_LOG_LEVEL);
+
+#define MCP4728_MAX_CHANNEL			4
+#define MCP4728_RESOLUTION			0xC
+#define MCP4728_DAC_MAX_VAL			((2 << (MCP4728_RESOLUTION - 1)) - 1)
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4728.c:18]: More easy to understand:
```suggestion
#define MCP4728_DAC_MAX_VAL			((1U << MCP4728_RESOLUTION) - 1U)
```

File: drivers/dac/dac_mcp4728.c [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+#define DT_DRV_COMPAT microchip_mcp4728
+
+#include <zephyr.h>
+#include <drivers/dac.h>
+#include <drivers/i2c.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(dac_mcp4728, CONFIG_DAC_LOG_LEVEL);
+
+#define MCP4728_MAX_CHANNEL			4
```
Comment (martinjaeger) [inline drivers/dac/dac_mcp4728.c:16]: Please add `U` suffix (e.g. `4U`) for all unsigned numbers (probably all).

**Review by martinjaeger — DISMISSED** [review]
Thanks for the contribution.

It looks like the similarities between this one and the existing MCP4725 are too small to justify a common driver, so I'm happy to go ahead like this.

Only got some minor style comments, which are not blocking.

**Review by martinjaeger — DISMISSED** [review]

File: drivers/dac/dac_mcp4728.c [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	.write_value = mcp4728_write_value,
+};
+
+#define INST_DT_MCP4728(index)							\
+	static const struct mcp4728_config mcp4728_config_##index = {		\
+		.bus = I2C_DT_SPEC_INST_GET(index),				\
+		.power_down = DT_INST_PROP(index, power_down_mode),		\
+		.voltage_reference = DT_INST_PROP(index, voltage_reference),	\
+		COND_CODE_1(DT_NODE_HAS_PROP(index, gain),			\
+			(.gain = DT_INST_PROP(index, gain)),			\
```
Comment (henrikbrixandersen) [inline drivers/dac/dac_mcp4728.c:102]: You can use `DT_INST_PROP_OR()`.

File: dts/bindings/dac/microchip,mcp4728.yaml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    gain:
+      type: array
+      required: false
+      default: [0, 0, 0, 0]
+      description: |
+        Gain selection bit.
+        0 = x1 (gain of 1)
+        1 = x2 (gain of 2)
+        Note: applicable only when internal Vref is selected
```
Comment (henrikbrixandersen) [inline dts/bindings/dac/microchip,mcp4728.yaml:38]: Perhaps add a note about the array entries corresponding to channels?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by martinjaeger — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**